File: BoardGameGeek/src/com/boardgamegeek/provider/GamesIdPollsNameResultsKeyProvider.java
package com.boardgamegeek.provider;

import android.database.sqlite.SQLiteDatabase;
import android.net.Uri;
import android.provider.BaseColumns;

import com.boardgamegeek.provider.BggContract.GamePollResults;
import com.boardgamegeek.provider.BggContract.GamePolls;
import com.boardgamegeek.provider.BggContract.Games;
import com.boardgamegeek.provider.BggDatabase.Tables;
import com.boardgamegeek.util.SelectionBuilder;

public class GamesIdPollsNameResultsKeyProvider extends BaseProvider {

	@Override
	protected SelectionBuilder buildExpandedSelection(Uri uri) {
		int gameId = Games.getGameId(uri);
		String pollName = Games.getPollName(uri);
		String players = Games.getPollResultsKey(uri);
		return new SelectionBuilder().table(Tables.POLLS_JOIN_POLL_RESULTS)
				.mapToTable(BaseColumns._ID, Tables.GAME_POLL_RESULTS).whereEquals(GamePolls.GAME_ID, gameId)
				.whereEquals(GamePolls.POLL_NAME, pollName).whereEquals(GamePollResults.POLL_RESULTS_PLAYERS, players);
	}

	@Override
	protected SelectionBuilder buildSimpleSelection(Uri uri) {
		int gameId = Games.getGameId(uri);
		String pollName = Games.getPollName(uri);
		String key = Games.getPollResultsKey(uri);
		return new SelectionBuilder()
				.table(Tables.GAME_POLL_RESULTS)
				.mapToTable(BaseColumns._ID, Tables.GAME_POLL_RESULTS)
				.where("poll_id = (SELECT game_polls._id FROM game_polls WHERE game_id=? AND poll_name=?)",
						String.valueOf(gameId), pollName).whereEquals(GamePollResults.POLL_RESULTS_PLAYERS, key);
	}

	@Override
	protected String getPath() {
		return "games/#/polls/*/results/*";
	}

	@Override
	protected String getType(Uri uri) {
		return GamePollResults.CONTENT_ITEM_TYPE;
	}

	@Override
	protected void deleteChildren(SQLiteDatabase db, SelectionBuilder builder) {
		new GamesIdPollsNameResultsProvider().deleteChildren(db, builder);
	}
}

File: app/src/main/java/com/boardgamegeek/provider/GamesIdPollsNameResultsKeyProvider.kt
package com.boardgamegeek.provider

import android.net.Uri
import android.provider.BaseColumns
import com.boardgamegeek.provider.BggContract.*
import com.boardgamegeek.provider.BggContract.Companion.PATH_GAMES
import com.boardgamegeek.provider.BggContract.Companion.PATH_POLLS
import com.boardgamegeek.provider.BggContract.Companion.PATH_POLL_RESULTS
import com.boardgamegeek.provider.BggDatabase.Tables

class GamesIdPollsNameResultsKeyProvider : BaseProvider() {
    override val path = "$PATH_GAMES/#/$PATH_POLLS/*/$PATH_POLL_RESULTS/*"

    override fun getType(uri: Uri) = GamePollResults.CONTENT_ITEM_TYPE

    public override fun buildSimpleSelection(uri: Uri): SelectionBuilder {
        val gameId = Games.getGameId(uri)
        val pollName = Games.getPollName(uri)
        val key = Games.getPollResultsKey(uri)
        return SelectionBuilder()
            .table(Tables.GAME_POLL_RESULTS)
            .mapToTable(BaseColumns._ID, Tables.GAME_POLL_RESULTS)
            .where(
                "${GamePollResults.Columns.POLL_ID} = (SELECT ${Tables.GAME_POLLS}.${BaseColumns._ID} FROM ${Tables.GAME_POLLS} WHERE ${GamePolls.Columns.GAME_ID}=? AND ${GamePolls.Columns.POLL_NAME}=?)",
                gameId.toString(),
                pollName
            )
            .whereEquals(GamePollResults.Columns.POLL_RESULTS_PLAYERS, key)
    }

    override fun buildExpandedSelection(uri: Uri): SelectionBuilder {
        val gameId = Games.getGameId(uri)
        val pollName = Games.getPollName(uri)
        val players = Games.getPollResultsKey(uri)
        return SelectionBuilder().table(Tables.POLLS_JOIN_POLL_RESULTS)
            .mapToTable(BaseColumns._ID, Tables.GAME_POLL_RESULTS)
            .whereEquals(GamePolls.Columns.GAME_ID, gameId)
            .whereEquals(GamePolls.Columns.POLL_NAME, pollName)
            .whereEquals(GamePollResults.Columns.POLL_RESULTS_PLAYERS, players)
    }
}
